File: umbra-go/sync.go
package umbra

import (
	"context"
	"encoding/json"
	"net/url"
	"strconv"
)

const SyncProtocolVersion uint32 = 1

type SyncClient struct {
	api *apiClient
}

type SyncSpace struct {
	Name string `json:"name"`
}

type SyncRecordKey struct {
	Namespace  string `json:"namespace"`
	Collection string `json:"collection"`
	RecordID   string `json:"record_id"`
}

type SyncOperation string

const (
	SyncOperationUpsert SyncOperation = "upsert"
	SyncOperationDelete SyncOperation = "delete"
)

type SyncMutation struct {
	MutationID    string          `json:"mutation_id"`
	Key           SyncRecordKey   `json:"key"`
	SchemaVersion uint32          `json:"schema_version"`
	BaseVersion   uint64          `json:"base_version"`
	Operation     SyncOperation   `json:"operation"`
	Payload       json.RawMessage `json:"payload,omitempty"`
}

type SyncExchangeInput struct {
	ProtocolVersion uint32         `json:"protocol_version"`
	Space           SyncSpace      `json:"space"`
	Cursor          string         `json:"cursor,omitempty"`
	Mutations       []SyncMutation `json:"mutations"`
	PullLimit       int            `json:"pull_limit,omitempty"`
}

type SyncAcceptedMutation struct {
	MutationID    string `json:"mutation_id"`
	RecordVersion uint64 `json:"record_version"`
	Cursor        string `json:"cursor"`
}

type SyncChange struct {
	Cursor         string          `json:"cursor"`
	Key            SyncRecordKey   `json:"key"`
	SchemaVersion  uint32          `json:"schema_version"`
	RecordVersion  uint64          `json:"record_version"`
	Operation      SyncOperation   `json:"operation"`
	Payload        json.RawMessage `json:"payload,omitempty"`
	WriterDeviceID string          `json:"writer_device_id"`
}

func (c SyncChange) DecodePayload(out any) error {
	return json.Unmarshal(c.Payload, out)
}

type SyncConflict struct {
	MutationID string      `json:"mutation_id"`
	Reason     string      `json:"reason"`
	Current    *SyncChange `json:"current,omitempty"`
}

type SyncRejectedMutation struct {
	MutationID string `json:"mutation_id"`
	Reason     string `json:"reason"`
}

type SyncExchangeResult struct {
	Accepted       []SyncAcceptedMutation `json:"accepted"`
	Conflicts      []SyncConflict         `json:"conflicts"`
	Rejected       []SyncRejectedMutation `json:"rejected"`
	Changes        []SyncChange           `json:"changes"`
	NextCursor     string                 `json:"next_cursor"`
	HasMore        bool                   `json:"has_more"`
	ResetRequired  bool                   `json:"reset_required"`
	Reason         string                 `json:"reason,omitempty"`
	SnapshotCursor string                 `json:"snapshot_cursor,omitempty"`
}

type SyncSnapshotInput struct {
	ProtocolVersion uint32
	SpaceName       string
	Cursor          string
	Limit           int
}

type SyncSnapshotPage struct {
	Records        []SyncChange `json:"records"`
	NextCursor     string       `json:"next_cursor,omitempty"`
	ExchangeCursor string       `json:"exchange_cursor"`
	HasMore        bool         `json:"has_more"`
}

func NewUpsertMutation(mutationID string, key SyncRecordKey, schemaVersion uint32, baseVersion uint64, payload any) (SyncMutation, error) {
	encoded, err := json.Marshal(payload)
	if err != nil {
		return SyncMutation{}, err
	}
	return SyncMutation{
		MutationID: mutationID, Key: key, SchemaVersion: schemaVersion, BaseVersion: baseVersion,
		Operation: SyncOperationUpsert, Payload: encoded,
	}, nil
}

func NewDeleteMutation(mutationID string, key SyncRecordKey, schemaVersion uint32, baseVersion uint64) SyncMutation {
	return SyncMutation{
		MutationID: mutationID, Key: key, SchemaVersion: schemaVersion, BaseVersion: baseVersion,
		Operation: SyncOperationDelete,
	}
}

func (s *SyncClient) Exchange(ctx context.Context, input SyncExchangeInput) (*SyncExchangeResult, error) {
	if input.ProtocolVersion == 0 {
		input.ProtocolVersion = SyncProtocolVersion
	}
	if input.Mutations == nil {
		input.Mutations = []SyncMutation{}
	}
	var out SyncExchangeResult
	if err := s.api.post(ctx, "/client/sync/exchange", input, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

func (s *SyncClient) Snapshot(ctx context.Context, input SyncSnapshotInput) (*SyncSnapshotPage, error) {
	if input.ProtocolVersion == 0 {
		input.ProtocolVersion = SyncProtocolVersion
	}
	query := url.Values{}
	query.Set("protocol_version", strconv.FormatUint(uint64(input.ProtocolVersion), 10))
	query.Set("space", input.SpaceName)
	if input.Cursor != "" {
		query.Set("cursor", input.Cursor)
	}
	if input.Limit > 0 {
		query.Set("limit", strconv.Itoa(input.Limit))
	}
	var out SyncSnapshotPage
	if err := s.api.get(ctx, "/client/sync/snapshot", query, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

File: umbra-typescript/src/sync.ts
import type { ApiClient } from './api'

export const syncProtocolVersion = 1

export type JsonPrimitive = string | number | boolean | null
export type JsonValue = JsonPrimitive | JsonValue[] | { [key: string]: JsonValue }

export interface SyncSpace {
  name: string
}

export interface SyncRecordKey {
  namespace: string
  collection: string
  record_id: string
}

export type SyncOperation = 'upsert' | 'delete'

export interface SyncMutation<T extends JsonValue = JsonValue> {
  mutation_id: string
  key: SyncRecordKey
  schema_version: number
  base_version: number
  operation: SyncOperation
  payload?: T
}

export interface SyncExchangeInput<T extends JsonValue = JsonValue> {
  protocol_version?: number
  space: SyncSpace
  cursor?: string
  mutations?: SyncMutation<T>[]
  pull_limit?: number
}

export interface SyncAcceptedMutation {
  mutation_id: string
  record_version: number
  cursor: string
}

export interface SyncChange<T extends JsonValue = JsonValue> {
  cursor: string
  key: SyncRecordKey
  schema_version: number
  record_version: number
  operation: SyncOperation
  payload?: T
  writer_device_id: string
}

export interface SyncConflict<T extends JsonValue = JsonValue> {
  mutation_id: string
  reason: string
  current?: SyncChange<T>
}

export interface SyncRejectedMutation {
  mutation_id: string
  reason: string
}

export interface SyncExchangeResult<T extends JsonValue = JsonValue> {
  accepted: SyncAcceptedMutation[]
  conflicts: SyncConflict<T>[]
  rejected: SyncRejectedMutation[]
  changes: SyncChange<T>[]
  next_cursor: string
  has_more: boolean
  reset_required: boolean
  reason?: string
  snapshot_cursor?: string
}

export interface SyncSnapshotInput {
  protocolVersion?: number
  spaceName: string
  cursor?: string
  limit?: number
}

export interface SyncSnapshotPage<T extends JsonValue = JsonValue> {
  records: SyncChange<T>[]
  next_cursor?: string
  exchange_cursor: string
  has_more: boolean
}

export function newUpsertMutation<T extends JsonValue>(
  mutationId: string,
  key: SyncRecordKey,
  schemaVersion: number,
  baseVersion: number,
  payload: T,
): SyncMutation<T> {
  return {
    mutation_id: mutationId,
    key,
    schema_version: schemaVersion,
    base_version: baseVersion,
    operation: 'upsert',
    payload,
  }
}

export function newDeleteMutation(
  mutationId: string,
  key: SyncRecordKey,
  schemaVersion: number,
  baseVersion: number,
): SyncMutation {
  return {
    mutation_id: mutationId,
    key,
    schema_version: schemaVersion,
    base_version: baseVersion,
    operation: 'delete',
  }
}

export class SyncClient {
  readonly #api: ApiClient

  constructor(api: ApiClient) {
    this.#api = api
  }

  exchange<T extends JsonValue = JsonValue>(input: SyncExchangeInput<T>) {
    return this.#api.post<SyncExchangeResult<T>>('/client/sync/exchange', {
      protocol_version: input.protocol_version || syncProtocolVersion,
      space: input.space,
      cursor: input.cursor || '',
      mutations: input.mutations || [],
      ...(input.pull_limit ? { pull_limit: input.pull_limit } : {}),
    })
  }

  snapshot<T extends JsonValue = JsonValue>(input: SyncSnapshotInput) {
    const query = new URLSearchParams()
    query.set('protocol_version', String(input.protocolVersion || syncProtocolVersion))
    query.set('space', input.spaceName)
    if (input.cursor) query.set('cursor', input.cursor)
    if (input.limit) query.set('limit', String(input.limit))
    return this.#api.get<SyncSnapshotPage<T>>('/client/sync/snapshot', query)
  }
}
